DESIGN.md

```diff
@@ -0,0 +1,387 @@
+# COSE Golang Library
+
+## Objectives
+
+The objectives of the `go-cose` library of this doc are
+
+- Implement `COSE_Sign` and `COSE_Sign1` as specified in [RFC8152](https://datatracker.ietf.org/doc/html/rfc8152).
+- Make the implementation secure and efficient.
+- Provide signing and verification based on golang built-in crypto libraries.
+- Provide extensibility so that other implementation can be used for signing and verification.
+  - This objective implies remote signing.

```

Comment:
Here's how we've defined it in Notary: https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md
With an example for Azure Key Vault: https://github.com/Azure/notation-azure-kv/blob/main/docs/nv2-sign-verify-aks.md

Reply:
> Here's how we've defined it in Notary: https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md
> With an example for Azure Key Vault: https://github.com/Azure/notation-azure-kv/blob/main/docs/nv2-sign-verify-aks.md

Thanks for the pointers, very useful (for me).

Digging around a bit more, I found [this](https://github.com/jwx-go/crypto-signer), which contains a couple of adapters to AWS and GCP's KMS for the [jwx](https://github.com/lestrrat-go/jwx) package.  Note that they are implementations of the `crypto.Signer` interface.  If we decoupled based on the same abstraction we could reuse them as-is from JOSE to COSE :-)

Reply:
Nice, @shizhMSFT, what do you think?

Reply:
We should not take the `crypto.Signer` interface.

According to the [godoc](https://pkg.go.dev/crypto#Signer) of the `crypto.Signer` interface,

> Sign signs digest with the private key, possibly using entropy from
> rand. For an RSA key, the resulting signature should be either a
> PKCS #<span/>1 v1.5 or PSS signature (as indicated by opts). For an (EC)DSA
> key, it should be a DER-serialised, ASN.1 signature structure.

For RSA, the resulted signature is ambiguous since it can be **PKCS v1.5** or **PSS**, and [*crypto.PSSOptions](https://pkg.go.dev/crypto/rsa#PSSOptions) needs to be passed in for **PSS**. However, this option is implementation specific. Therefore, the `opts` is redundant for most implementation, and we need special handling for build-in golang RSA implementation.

For ECDSA, it is even worse. `Sign` returns ASN.1 DER-encoded signature, which is incompatible with COSE since COSE requires `I2OSP` encoding.

Here's the ASN.1 notation for ECDSA signatures.
```
ECDSASignature ::= SEQUENCE {
  r INTEGER
  s INTEGER
}
```
If we adopt `crypto.Signer`, we need to convert all ASN.1 DER-encoded signature from / to COSE signatures, which requires extra ASN.1 encoder / decoder and extra checks. _COSE is from CBOR. We should avoid ASN.1 in the first place._

Another important thing is that the `crypto` package has `Signer` interface defined but there is no `Verifier` interface. This is not a good sign as implementations can have various interfaces.

Of course, we can always build adapters for those existing implementations.

Reply:
I tend to agree with your assessment about the troubles around adapting crypto.Signer.  However, imperfect as it is, that's what the standard package offers, and consequently what the ecosystem has been building against in the past few years.  I have found quite a lot of implementations of the Signer, for both cloud KMS and hardware HSMs, and that is all good stuff that can be reused as-is if we decided to adopt the same interface.

So, to me the choice is not super easy because of all the advantages we'd get from going in the "standards" direction vs the extra care we need to exert when coding around the wobbly parts of the interface.

PS: I have looked at other places too, for example the [CIRCL library](https://github.com/cloudflare/circl/blob/master/sign/sign.go) which curiously builds its PrivateKey interface as a superset of crypto.Signer (probably for backwards compat reasons?)

Reply:
Guess what? We can make `NewSigner` be

```go
func NewSigner(alg *Algorithm, key crypto.Signer) (Signer, error)
```

so that we can make it compatible with all previous implementations.

Reply:
👍